cdap-app-fabric/src/main/java/co/cask/cdap/app/runtime/AbstractProgramRuntimeService.java

```diff
@@ -71,12 +84,53 @@ public RuntimeInfo run(Program program, ProgramOptions options) {
     ProgramRunner runner = programRunnerFactory.create(ProgramRunnerFactory.Type.valueOf(program.getType().name()));
     Preconditions.checkNotNull(runner, "Fail to get ProgramRunner for type " + program.getType());
     ProgramOptions optionsWithRunId = addRunId(options, RunIds.generate());
-    final RuntimeInfo runtimeInfo = createRuntimeInfo(runner.run(program, optionsWithRunId), program);
+    ProgramOptions optionsWithPlugins = addPluginLocations(optionsWithRunId,
+                                                           Id.Namespace.from(program.getNamespaceId()),
+                                                           program.getApplicationSpecification());
+    final RuntimeInfo runtimeInfo = createRuntimeInfo(runner.run(program, optionsWithPlugins), program);
     programStarted(runtimeInfo);
     return runtimeInfo;
   }
 
   /**
+   * Return the copy of the {@link ProgramOptions} including locations of plugin artifacts in it.
+   * @param options the {@link ProgramOptions} in which the locations of plugin artifacts needs to be included
+   * @param namespace namespace of the Program
+   * @param appSpec program's Application Specification
+   * @return the copy of the program options with locations of plugin artifacts included in them
+   */
+  private ProgramOptions addPluginLocations(ProgramOptions options, Id.Namespace namespace,
+                                            @Nullable ApplicationSpecification appSpec) {
+    // appSpec is null in an unit test
+    if (appSpec == null) {
+      return options;
+    }
+
+    Id.Artifact appArtifactId = Id.Artifact.from(namespace, appSpec.getArtifactId());
+    ImmutableMap.Builder<String, String> builder = ImmutableMap.builder();
+    builder.putAll(options.getArguments().asMap());
+    Map<String, String> artifactLocations = Maps.newHashMap();
+    for (Map.Entry<String, Plugin> pluginEntry : appSpec.getPlugins().entrySet()) {
+      Plugin plugin = pluginEntry.getValue();
+      Map.Entry<ArtifactDescriptor, PluginClass> artifactEntry;
+      try {
+        artifactEntry = artifactRepository.getPlugin(
+          appArtifactId, plugin.getPluginClass().getType(), plugin.getPluginClass().getName(), plugin.getArtifactId());
+      } catch (IOException e) {
+        throw Throwables.propagate(e);
+      } catch (PluginNotExistsException e) {
+        throw new IllegalArgumentException(String.format("Plugin of type %s, name %s could not be found",
+                                                         plugin.getPluginClass().getType(),
+                                                         plugin.getPluginClass().getName()), e);
+      }
+      artifactLocations.put(plugin.getArtifactId().toString(), artifactEntry.getKey().getLocation().getName());
```

Comment:
Move this into the try block. I like statement closer to the context. Also, the declaration of the artifactEntry can be moved inside the try block as well.